hudi-client/hudi-spark-client/src/test/java/org/apache/hudi/table/TestHoodieSparkTable.java

```diff
@@ -0,0 +1,137 @@
+    additionalFiles.forEach(fileName -> {
+      markerList.add(partitionPath + "/" + fileName);
+    });
+
+    HoodieWriteConfig writeConfig = HoodieWriteConfig.newBuilder().withPath(basePath).withMarkersType(MarkerType.DIRECT.name()).build();
+    when(writeMarkers.doesMarkerDirExist()).thenReturn(true);
+    when(writeMarkers.createdAndMergedDataPaths(getEngineContext(), writeConfig.getFinalizeWriteParallelism())).thenReturn(markerList);
+
+    HoodieTableMetaClient metaClient = mock(HoodieTableMetaClient.class);
+    when(metaClient.getBasePath()).thenReturn(new StoragePath(basePath));
+    when(metaClient.getTableType()).thenReturn(HoodieTableType.COPY_ON_WRITE);
+    HoodieStorage storage = mock(HoodieStorage.class);
+    when(metaClient.getStorage()).thenReturn(storage);
+
+    additionalFiles.forEach(fileName -> {
+      try {
+        StoragePath storagePath = new StoragePath(basePath + "/" + partitionPath + "/" + fileName);
+        if (failureType == DeleteFailureType.TRUE_ON_DELETE) {
+          when(storage.deleteFile(storagePath)).thenReturn(true);
+        } else if (failureType == DeleteFailureType.FALSE_ON_DELETE_IS_EXISTS_FALSE) {
+          when(storage.deleteFile(storagePath)).thenReturn(false);
+          when(storage.exists(storagePath)).thenReturn(false);
+        } else if (failureType == DeleteFailureType.FALSE_ON_DELETE_IS_EXISTS_TRUE) {
+          when(storage.deleteFile(storagePath)).thenReturn(false);
+          when(storage.exists(storagePath)).thenReturn(true);
+        } else if (failureType == DeleteFailureType.FILE_NOT_FOUND_EXC_ON_DELETE) {
+          when(storage.deleteFile(storagePath)).thenThrow(new FileNotFoundException("throwing file not found exception"));
+        } else {
+          // run time exception
+          when(storage.deleteFile(storagePath)).thenThrow(new RuntimeException("throwing run time exception"));
+        }
+        // lets create the data file. so that we can validate later.
+        localStorage.create(storagePath);
+      } catch (IOException e) {
+        throw new HoodieException("Failed to check data file existance " + fileName);
+      }
+    });
+    HoodieTable hoodieTable = HoodieSparkTable.create(writeConfig, getEngineContext(), metaClient);
+    if (failureType == DeleteFailureType.RUNTIME_EXC_ON_DELETE || failureType == DeleteFailureType.FALSE_ON_DELETE_IS_EXISTS_TRUE) {
+      assertThrows(HoodieException.class, () -> {
+        hoodieTable.reconcileAgainstMarkers(getEngineContext(), "0001", writeStatList, false, false, writeMarkers);
+      });
+    } else { // all other cases
+      hoodieTable.reconcileAgainstMarkers(getEngineContext(), "0001", writeStatList, false, false, writeMarkers);
+      // validate that additional files are deleted from storage
+      additionalFiles.forEach(fileName -> {
+        try {
+          verify(storage, times(1)).deleteFile(new StoragePath(basePath + "/" + partitionPath + "/" + fileName));
```

Comment:
Is it possible to have more granular assertion to show file got deleted. I guess this is sufficient as we are working with mocks in the test, but just a thought from future-proofing pov.

Reply:
I felt mock is sufficient here. don't wanna add a functional test for this.